README
Liverpie 0.5
~~~~~~~~~~~~

Liverpie is a Language Independent IVR Proxy. It stays between Freeswitch
and any web app, written in any language and running on any platform, that
functions as a state machine that describes an IVR (Interactive Voice
Response system).

There are many APIs and software products available that will help you write
an IVR using Freeswitch, but they all depend on a particular programming 
language and require that you either learn that language, or at least the
API itself.

With Liverpie, you can write your IVR in your favourite language, as a
regular web app, as long as it runs like a state machine (you keep calling
the same thing and it gives different replies each time, depending on a
predefined schema). You can code that in any way you like, and there are
a few examples on the site.

Liverpie will talk to your web app on one end, and to Freeswitch on the 
other -- that's what qualifies it as a proxy.

Full documentation is now only maintained at http://www.liverpie.com because 
I got tired of editing the same things in two places.


CREDITS
-------
Jonathan Palley, author of Telegraph for FreeSWITCH, wrote the part that
instantiates an EventMachine and creates a server ready for FreeSWITCH.
I wrote everything else and changed some of Jonathan's code.

    Contact me at: alxx@indigenious.ro
            Yahoo Messenger: alex_deva2002
            Skype: alex_deva2002
            ICQ: 2723153
            Google Talk: alexandru.deva
            Phone: +4 0722 878 295.

I'm available for consultancy regarding installation and usage of Liverpie.

Adeel Ansari helped with the Java example and also worked with me to fix
a bug that prevented some Java webservers that don't always send a cookie
to work with Liverpie.


11. LICENSE
-----------
The MIT License

Copyright (c) 2008 Alex Deva, parts of code written by Jonathan Palley

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG
17 apr 2018 - version 0.6
 * Gemfile

26 feb 2009 - version 0.5
 * Liverpie now can handle multiple web applications! Run several IVRs with the same Liverpie instance.
 * Changed syntax of config file accordingly; added convert script.
 * Fixed some bugs, cleaned up a bit.

1 feb 2009 - version 0.4
 * fixed an issue where Liverpie ignored parameters received in two batches, therefore causing problems with call_direction
 * fixed an issue where Liverpie could set nil cookies

14 nov 2008 - version 0.3
 * added support for DTMF notification (a separate webapp uri which gets called everytime there's a DTMF event with the respective value)
 * fixed bug with generating webapp parameters

25 sept 2008 - version 0.2
 * fixed a bug that prevented web servers that don't send cookies with each request (like Tomcat) to work with Liverpie
 * fixed a bug where Liverpie would crash if the webapp became unavailable. Now it just carelessly logs it.


lib/script_base.rb
require 'optparse'
require 'ostruct'
require 'pp'
require 'rubygems'

module Liverpie
  class ScriptOptions
    def self.parse(args, process)
      options=OpenStruct.new
      options.daemonize = false
      options.pid_file = "log/liverpie.pid"
      options.cwd = Dir.pwd
      options.action = 'start'
      options.debug = false
      options.verbose = false
      opts = OptionParser.new do |opts|
        opts.banner = "Usage: #{process} [options]"
        opts.separator  ""
        opts.separator "Specific options:"

        opts.on('-d', '--daemon', 'Run as daemon', 'runs in shell by default') do |ext|
          options.daemonize=true
        end
        opts.on('-p','--pid FILE', 'PID File', 'Defaults to log directory') do |ext|
           options.pid_file = ext
        end
        opts.on('-a', '--action [TYPE]', [:start, :stop], "Action To Do") do |action|
          options.action = action
        end
        opts.on('-C PATH', 'Change dir before starting') do |path|
          options.cwd = path
        end
        opts.on('-v', '--verbose', 'Log communication') do |verbose|
          options.verbose = true
        end
        opts.on('-l', '--list', 'Answer to all events') do |debug|
          options.debug = true
        end
      end

      opts.parse!(args)
      options
    end
  end

  class ScriptHandler
     def initialize(process, &block)
       @process_name = process
       @options=ScriptOptions.parse(ARGV, process)
       @pid_file =File.join(@options.cwd, @options.pid_file)
       if @options.action == 'start'
          self.start &block
       else
         self.stop
        end
     end

     def start    
      log "Starting Server"

       @log_file = File.expand_path(File.join(File.dirname(__FILE__), "../log/liverpie.log"))
        
        log "Logging at #{@log_file}"

        if @options.daemonize
           if File.exists?(@pid_file)
              log "PID File Already Exists!  Exiting..."
              exit
            else
              daemonize
              write_pid_file
            end
         end
        setup_signals

        # require 'commands/servers/base'

        # ActiveRecord::Base.allow_concurrency = true #Prevents connection from dieing... (alxx: what connection?)
        require 'eventmachine'
        
        # tail_thread = tail(Pathname.new(@log_file).cleanpath) unless @options.daemonize # alxx: what's this?
      
       
         begin
            yield @options.verbose, @options.debug
         ensure
          # tail_thread.kill if tail_thread #alxx: hope we don't need this.
           log 'Exiting'
         end
    
 
     end
     def stop
      send_signal("INT", @pid_file)    
     end

     def kill
       log "Exiting..."
       File.unlink(@pid_file) if @pid_file and File.exists?(@pid_file)
      #Need to do this twice as voice_events catches it...
      exit!
     end

     def setup_signals
     # forced shutdown, even if previously restarted (actually just like TERM but for CTRL-C)
         # clean up the pid file always
        at_exit { 
          log "Exiting..."
          File.unlink(@pid_file) if @pid_file and File.exists?(@pid_file) 
          }
      
           trap("INT") { log "INT signal received."; kill}
          if RUBY_PLATFORM !~ /mswin/
            # graceful shutdown
            trap("TERM") { log "TERM signal received."; kill }
            log "Signals ready.  TERM => stop. INT => stop (no restart)."
          end
   
   
     end
     def send_signal(signal, pid_file)
     begin 
       pid = open(pid_file).read.to_i
       print "Sending #{signal} to #{@options.process} at PID #{pid}..."
       Process.kill(signal, pid)
     rescue Errno::ESRCH
       print "Process does not exist.  Not running."
     rescue Errno::ENOENT
        print "No .pid file.  Not shutting down"
      end
  
      puts "Done."
      end
     def log(msg)
 #      STDERR.print "** ", msg, " **\n"
        puts "** #{msg} **"
     end
   
    # Writes the PID file but only if we're on windows.
     def write_pid_file
       log "Write pid"
       log @pid_file
     
       if RUBY_PLATFORM !~ /mswin/
          open(@pid_file,"w") {|f| f.write(Process.pid) }
       end
     end
    def daemonize(options={})
       #ops = resolve_defaults(options)
        # save this for later since daemonize will hose it
        if RUBY_PLATFORM !~ /mswin/
          require 'daemons/daemonize'
          log 'daemonizing!'
          Daemonize.daemonize(@log_file)

          #log File.join(@options.cwd, @options.pid_file)
          # change back to the original starting directory
          Dir.chdir(@options.cwd)
        else
          log "WARNING: Win32 does not support daemon mode."
        end
      end

  end
end

lib/liverpie_client.rb
require 'net/http'

class LiverpieClient

  def initialize
    webapps = Liverpie.webapps
    @webapp_connections = {}
    webapps.keys.each do |webapp_id| 
      @webapp_connections[webapp_id] = { 
                      :http => Net::HTTP.new(webapps[webapp_id]['ip'], webapps[webapp_id]['port']),
                      :state_machine_uri => webapps[webapp_id]['state_machine_uri'],
                      :dtmf_uri => webapps[webapp_id]['dtmf_uri'],
                      :reset_uri => webapps[webapp_id]['reset_uri']
                     }
    end
  end
  
  # Send a regular event triggered by the webapp. This is the state machine call.
  def send(params, cookie)
    requested_webapp_id = params['variable_liverpie_webapp_id']
    Liverpie.log "Calling webapp #{requested_webapp_id}..."
    
    hdrs = {}
    hdrs['Cookie'] = cookie if cookie
    
    webapp_connection = get_webapp_connection(requested_webapp_id)
    return [nil, nil] unless webapp_connection
    
    partxt = params.keys.map { |k| "#{k}=#{params[k]}"}*'&'
    
    begin
      resp = webapp_connection[:http].post(webapp_connection[:state_machine_uri], partxt, hdrs)
    rescue Exception => e
      Liverpie.log "Error - could not reach the webapp: #{e.message}"
      return [nil, nil]
    end
    
    case resp.code.to_i
    when 200, 302
      received_cookie = resp['Set-Cookie']
      Liverpie.log "Got webapp response for runner method"
    else
      Liverpie.log "Error - code(#{resp.code}) - #{resp.msg}"
    end
    
    return [(resp.code.to_i == 200 ? resp.body : nil), received_cookie.to_s.empty? ? cookie : received_cookie]
  end
  
  # Asynchronously notify the webapp of all the inbound DTMF codes.
  def send_dtmf(code, cookie, requested_webapp_id)
    Liverpie.log "Sending DTMF #{code} to webapp #{requested_webapp_id}..."
    
    webapp_connection = get_webapp_connection(requested_webapp_id)
    return nil, cookie unless webapp_connection
    
    # If this wasn't configured, then ignore DTMF
    return nil, cookie if webapp_connection[:dtmf_uri].to_s.empty?
        
    hdrs = {}
    hdrs['Cookie'] = cookie if cookie
    
    begin
      resp = webapp_connection[:http].post(webapp_connection[:dtmf_uri], "dtmf_code=#{code}", hdrs)
    rescue Exception => e
      Liverpie.log "Error - could not reach the webapp: #{e.message}"
      return [nil, nil]
    end
    
    case resp.code.to_i
    when 200, 302
      received_cookie = resp['Set-Cookie']
      Liverpie.log "Got webapp response for dtmf method"
    else
      Liverpie.log "Error - code(#{resp.code}) - #{resp.msg}"
    end
    
    return [(resp.code.to_i == 200 ? resp.body : nil), received_cookie.to_s.empty? ? cookie : received_cookie]
  end
  
  # Reset the webapp state machine
  def reset(requested_webapp_id)
    Liverpie.log "Resetting state machine in webapp #{requested_webapp_id}..."
    
    webapp_connection = get_webapp_connection(requested_webapp_id)
    return nil unless webapp_connection
    
    begin
      resp = webapp_connection[:http].get(webapp_connection[:reset_uri])    
    rescue Exception => e
      Liverpie.log "Error - could not reach the webapp: #{e.message}"
      return nil
    end
    
    case resp.code.to_i
    when 200, 302
      received_cookie = resp['Set-Cookie']
      Liverpie.log "Got webapp response for reset method"
    else
      Liverpie.log "Error - code(#{resp.code}) - #{resp.msg}"
    end
    return received_cookie
  end
  
  private
  
  # Find out what webapp do we need to talk to
  def get_webapp_connection(requested_webapp_id)
    if requested_webapp_id.to_s.empty?
      Liverpie.log "Error - Freeswitch didn't send us a liverpie_webapp_id, check that you set liverpie_webapp_id in your dialplan before calling Liverpie"
      return false
    end
    
    webapp_connection = @webapp_connections[requested_webapp_id]
    
    if webapp_connection.nil?
      Liverpie.log "Error - No webapp named #{requested_webapp_id} is configured in liverpie.conf"
      return false
    end
    
    webapp_connection
  end
  
end

lib/liverpie_server.rb
require 'eventmachine'

module LiverpieServer
  def post_init
    Liverpie.log "New Connection"
    @receive_buffer = ''
    @connection_params = nil
    @initial_commands = []
    #@initial_commands << "log DEBUG\n\n"
    @initial_commands << "myevents\n\n"

    @initialized = false
    @latest_variables = {}
    @cookie = nil
    @client = LiverpieClient.new
    @expected_key = 'content_type'
    @expected_value = 'command/reply'
    
    send "connect"
  end
 
  def set_modes(verbose, debug)
    Liverpie.log "Turning on Debug Mode" if @debug = debug
    Liverpie.log "Turning on Verbose Mode" if @verbose = verbose
  end
  
  def receive_data data
    @receive_buffer << data
          
    # Act for data that ends in two newlines
    if @receive_buffer[-2..-1] == "\n\n"
      @params = hashify @receive_buffer
      @receive_buffer = ''
      
      Liverpie.log "Data Received (content-type: #{@params['content_type']}):\n#{data}" if @verbose

      # First we run the initial commands.
      unless @initialized
        
        Liverpie.log "Initializing connection..."

        @connection_params ||= @params
        cmd = @initial_commands.shift
        if cmd.is_a?(Hash)
          send_application cmd[:app], cmd[:params]
        else
          send cmd
        end
        @initialized = @initial_commands.empty?
        @latest_variables.merge!(@params)

        # Start the state machine also
        @cookie = @client.reset(@params['variable_liverpie_webapp_id']) if @initialized
      else   

        content_types = [@latest_variables['content_type'], @params['content_type']]
        if content_types - [ 'text/event-plain' ] != content_types # only merge we got one of those content_types
          @latest_variables.merge!(@params)
        end

        Liverpie.log "Waiting for params[#{@expected_key}] to be #{@expected_value}." if @verbose
        if @params[@expected_key] == @expected_value          
          resp, @cookie = @client.send(@latest_variables, @cookie)
          if resp.to_s.empty?
            Liverpie.log "PROBLEM: Didn't get any response from webapp"
          else
            yaml = YAML.load(resp)
            @expected_key, @expected_value = yaml['expected_key'], yaml['expected_value']
            send yaml['msg']
          end
        end
        
        # Liverpie.log "N Param keys: #{hashify(data).keys.inspect}"
        
        # If we find an inband DTMF code, always send it to the webapp.
        # In this case, the webapp may return an empty response, or an YAML. We don't do anything in the first case.
        if @params['dtmf_digit']
          resp, @cookie = @client.send_dtmf(@params['dtmf_digit'], @cookie, @latest_variables['variable_liverpie_webapp_id'])
          yaml = YAML.load(resp)
          msg  = yaml ? yaml['msg'].nil? ? nil : yaml['msg'] : nil
          send msg if msg
        end
        
        @started = true unless @started
 
      end # unless initialized
    end # data ends
  end

  def unbind
    puts "Disconnected from Freeswitch"
  end
 
  def hashify data
    hsh = Hash.new
    data.split("\n").each do |line|
      hsh[line.split(':')[0].strip.gsub('-', '_').downcase] = line.split(':')[1].strip unless line.empty?
    end
    hsh
  end

  def send_application app, params=nil
    msg = "SendMsg #{@connection_params['unique-id'].to_s}\ncall-command: execute\nexecute-app-name: #{app}"
    msg << "\nexecute-app-arg: #{params}" #if params
    send msg
  end

  def send data
    Liverpie.log "Sending:\n#{data}" if @verbose
    send_data data + "\n\n"
   end
end

config/convert_to_multiple_webapps.rb
#/usr/bin/env ruby

# This script will take your existing config.yml file (up to and including version 0.4)
# and convert it into the 0.5 format, which allows for multiple webapps to be run using the same Liverpie instance.
#

require 'yaml'

old_yml_file = 'liverpie.yml'

unless File.exist?(old_yml_file)
  puts "Could not find liverpie.yml. Please run this converter from the config directory"
  exit
end

old_yml = YAML.load_file old_yml_file

puts "Converting your configuration file. The new file will have an application named 'my_webapp' with your existing application data. Please change that as you wish."

config = old_yml['configuration']
new_yml_hash = {
  'configuration' => {
    'bind_ip'   => config['bind_ip'],
    'bind_port' => config['bind_port']
  },
  'webapps' => {
    'my_webapp' => {
      'ip'                => config['webapp_ip'],
      'port'              => config['webapp_port'],
      'state_machine_uri' => config['webapp_uri'],
      'reset_uri'         => config['webapp_reset_uri'],
      'dtmf_uri'          => config['webapp_dtmf_uri']
    }
  }
}

puts 'Renaming your configuration file into liverpie.yml.obsolete...'

File.rename old_yml_file, 'liverpie.yml.obsolete'

puts 'Saving converted configuration file...'

File.open('liverpie.yml', 'w') do |out|
  YAML.dump new_yml_hash, out
end

puts 'Done. Thank you!'
